domain/resource/state/resource.go

```diff
@@ -41,6 +41,163 @@ func NewState(factory database.TxnRunnerFactory, clock clock.Clock, logger logge
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to check there is no related resources in unit_resource.
+	noUnitResourceStmt, err := st.Prepare(`
+SELECT resource_uuid AS &resourceIdentity.uuid 
+FROM unit_resource
+WHERE resource_uuid IN ($uuids[:])`, resourceIdentity{}, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from resource_retrieved_by.
+	deleteFromRetrievedByStmt, err := st.Prepare(`
+DELETE FROM resource_retrieved_by
+WHERE resource_uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from application_resource.
+	deleteFromApplicationResourceStmt, err := st.Prepare(`
+DELETE FROM application_resource
+WHERE resource_uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from resource.
+	deleteFromResourceStmt, err := st.Prepare(`
+DELETE FROM resource
+WHERE uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) (err error) {
+		// list all resources for an application.
+		var resources []resourceIdentity
+		err = tx.Query(ctx, listAppResourcesStmt, appIdentity).GetAll(&resources)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return err
+		}
+		resUUIDs := make(uuids, 0, len(resources))
+		for _, res := range resources {
+			resUUIDs = append(resUUIDs, res.UUID)
+		}
+
+		checkLink := func(message string, stmt *sqlair.Statement) error {
+			var resources []resourceIdentity
+			err := tx.Query(ctx, stmt, resUUIDs).GetAll(&resources)
+			switch {
+			case errors.Is(err, sqlair.ErrNoRows): // Happy path
+				return nil
+			case err != nil:
+				return err
+			}
+			return errors.Errorf("%s: %w", message, resourceerrors.InvalidCleanUpState)
+		}
+
+		// check there are no related resources in resource_file_store.
+		if err = checkLink("resource linked to file store data", noFileStoreStmt); err != nil {
+			return errors.Capture(err)
+		}
+
+		// check there are no related resources in resource_image_store.
+		if err = checkLink("resource linked to image store data", noImageStoreStmt); err != nil {
+			return errors.Capture(err)
+		}
+
+		// check there are no related resources in unit_resource.
+		if err = checkLink("resource linked to unit", noUnitResourceStmt); err != nil {
+			return errors.Capture(err)
+		}
+
+		// delete resources from resource_retrieved_by.
+		if err = tx.Query(ctx, deleteFromRetrievedByStmt, resUUIDs).Run(); err != nil {
+			return errors.Capture(err)
+		}
+
+		// delete resources from application_resource.
+		if err = tx.Query(ctx, deleteFromApplicationResourceStmt, resUUIDs).Run(); err != nil {
+			return errors.Capture(err)
+		}
+
+		// delete resources from resource.
+		return tx.Query(ctx, deleteFromResourceStmt, resUUIDs).Run()
```

Comment:
Just occured to me, there should probably be a check on the outcome of this query to make sure that the resource was actually deleted. i.e. that the number of rows affected was 1. Same for `deleteFromApplicationResourceStmt` and for the unit deletion.

Reply:
Nice spot thanks !